File: javascript/leetcode75/array-strings/238. Product of Array Except Self.js
// https://leetcode.com/problems/product-of-array-except-self/
// Input: nums = [1,2,3,4]
// Output: [24,12,8,6]
class Solution {
    productExceptSelf(nums) {
        let l = nums.length;
        let products = new Array(l).fill(1);

        let left = 1;
        for (let i = 1; i < l; i++) {
            products[i] = nums[i - 1] * left;
            left = products[i];
        }
        
        console.log(products);
        let right = 1;
        for (let i = l - 1; i >= 0; i--) {
            products[i] *= right;
            right *= nums[i];
        }
        
        return products;
    }
}

let s = new Solution();
console.log(s.productExceptSelf([2, 3, 5, 0])); 

File: python/patterns/array-strings/238. Product of Array Except Self.py
# https://leetcode.com/problems/product-of-array-except-self/
# Input: nums = [1,2,3,4]
# Output: [24,12,8,6]

from typing import List

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        result = [1] * n
        
        # Calculate prefix products
        for i in range(1, n):
            result[i] = result[i-1] * nums[i-1]
        
        # Calculate suffix products and combine
        suffix = 1
        for i in range(n-1, -1, -1):
            result[i] = result[i] * suffix
            suffix *= nums[i]
        
        return result

s = Solution()
print(s.productExceptSelf([1,2,3,4]))

"""
Time complexity: O(N)
Space complexity: O(1) - excluding the output array
where N is the length of the input array.
""" 